Flaky test on engine-gecko-nightly GeckoEngineSessionTest.kt:272

This test fun `saveState()` have been failing intermittently  GeckoEngineSessionTest.kt:272
```
:support-base:packageDebugAssets
:support-ktx:mergeDebugShaders
:support-ktx:compileDebugShaders
:support-ktx:generateDebugAssets
:support-ktx:packageDebugAssets
:support-utils:mergeDebugShaders
:support-utils:compileDebugShaders
:support-utils:generateDebugAssets
:support-utils:packageDebugAssets
:browser-engine-gecko-nightly:mergeDebugAssets
:browser-engine-gecko-nightly:generateDebugUnitTestConfig
:browser-engine-gecko-nightly:mockableAndroidJar
:browser-engine-gecko-nightly:processDebugUnitTestJavaRes
:browser-engine-gecko-nightly:transformClassesAndResourcesWithPrepareIntermediateJarsForDebug
:browser-engine-gecko-nightly:testDebugUnitTest

mozilla.components.browser.engine.gecko.GeckoEngineSessionTest > saveState FAILED
    java.lang.UnsatisfiedLinkError at GeckoEngineSessionTest.kt:272
42 tests completed, 1 failed
:browser-engine-gecko-nightly:testDebugUnitTest FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':browser-engine-gecko-nightly:testDebugUnitTest'.
> There were failing tests. See the report at: file:///build/android-components/components/browser/engine-gecko-nightly/build/reports/tests/testDebugUnitTest/index.html
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
BUILD FAILED in 5m 47s
265 actionable tasks: 261 executed, 4 up-to-date
[taskcluster 2018-10-17 14:29:09.572Z] === Task Finished ===
[taskcluster 2018-10-17 14:29:09.572Z] Unsuccessful task run with exit code: 1 completed in 559.804 seconds
```

The same test failing in this PR https://github.com/mozilla-mobile/android-components/pull/1068

```
mozilla.components.browser.engine.gecko.GeckoEngineSessionTest > saveState FAILED
    java.lang.UnsatisfiedLinkError at GeckoEngineSessionTest.kt:273
45 tests completed, 1 failed
:browser-engine-gecko-nightly:testReleaseUnitTest FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':browser-engine-gecko-nightly:testReleaseUnitTest'.
> There were failing tests. See the report at: file:///build/android-components/components/browser/engine-gecko-nightly/build/reports/tests/testReleaseUnitTest/index.html
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
BUILD FAILED in 6m 21s
300 actionable tasks: 295 executed, 5 up-to-date
[taskcluster 2018-10-17 22:30:17.127Z] === Task Finished ===
[taskcluster 2018-10-17 22:30:17.127Z] Unsuccessful task run with exit code: 1 completed in 393.038 seconds

```

@pocmo  our favorite friend is here now in beta flavor :stuck_out_tongue_winking_eye:
https://github.com/mozilla-mobile/android-components/pull/1162

```
> Task :support-utils:compileDebugShaders
> Task :support-utils:generateDebugAssets
> Task :support-utils:packageDebugAssets
> Task :browser-engine-gecko-beta:mergeDebugAssets
> Task :browser-engine-gecko-beta:generateDebugUnitTestConfig
> Task :browser-engine-gecko-beta:mockableAndroidJar
> Task :browser-engine-gecko-beta:processDebugUnitTestJavaRes
> Task :browser-engine-gecko-beta:transformClassesAndResourcesWithPrepareIntermediateJarsForDebug
> Task :browser-engine-gecko-beta:testDebugUnitTest
mozilla.components.browser.engine.gecko.GeckoEngineSessionTest > loadUrl FAILED
    java.lang.UnsatisfiedLinkError at GeckoEngineSessionTest.kt:162
46 tests completed, 1 failed
> Task :browser-engine-gecko-beta:testDebugUnitTest FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':browser-engine-gecko-beta:testDebugUnitTest'.
> There were failing tests. See the report at: file:///build/android-components/components/browser/engine-gecko-beta/build/reports/tests/testDebugUnitTest/index.html
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
BUILD FAILED in 5m 43s
267 actionable tasks: 267 executed
[taskcluster 2018-10-24 17:41:24.348Z] === Task Finished ===
[taskcluster 2018-10-24 17:41:24.348Z] Unsuccessful task run with exit code: 1 completed in 349.182 seconds

```

--- a/components/browser/engine-gecko-beta/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt
+++ b/components/browser/engine-gecko-beta/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt
@@ -5,6 +5,7 @@
 package mozilla.components.browser.engine.gecko
 
 import android.os.Handler
+import android.os.Message
 import mozilla.components.browser.errorpages.ErrorType
 import mozilla.components.concept.engine.DefaultSettings
 import mozilla.components.concept.engine.EngineSession
@@ -22,6 +23,7 @@ import org.junit.Assert.assertFalse
 import org.junit.Assert.assertNotNull
 import org.junit.Assert.assertNull
 import org.junit.Assert.assertTrue
+import org.junit.Before
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.ArgumentMatchers.any
@@ -54,6 +56,11 @@ import org.robolectric.RobolectricTestRunner
 
 @RunWith(RobolectricTestRunner::class)
 class GeckoEngineSessionTest {
+
+    @Before
+    fun setup() {
+        ThreadUtils.sGeckoHandler = null
+    }
 
     @Test
     fun engineSessionInitialization() {
@@ -262,13 +269,22 @@ class GeckoEngineSessionTest {
 
     @Test
     fun saveState() {
-        ThreadUtils.sGeckoHandler = Handler()
-
         val engineSession = GeckoEngineSession(mock(GeckoRuntime::class.java))
         engineSession.geckoSession = mock(GeckoSession::class.java)
         val currentState = GeckoSession.SessionState("")
         val stateMap = mapOf(GeckoEngineSession.GECKO_STATE_KEY to currentState.toString())
 
+        var first = true
+        ThreadUtils.sGeckoHandler = object : Handler() {
+            override fun sendMessageAtTime(msg: Message?, uptimeMillis: Long): Boolean {
+                // Only run the saveState runnable and ignore everything that comes after
+                if (first) {
+                    first = false
+                    return super.sendMessageAtTime(msg, uptimeMillis)
+                }
+                return true
+            }
+        }
         `when`(engineSession.geckoSession.saveState()).thenReturn(GeckoResult.fromValue(currentState))
         assertEquals(stateMap, engineSession.saveState())
     }

--- a/components/browser/engine-gecko-nightly/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt
+++ b/components/browser/engine-gecko-nightly/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt
@@ -5,6 +5,7 @@
 package mozilla.components.browser.engine.gecko
 
 import android.os.Handler
+import android.os.Message
 import mozilla.components.browser.errorpages.ErrorType
 import mozilla.components.concept.engine.DefaultSettings
 import mozilla.components.concept.engine.EngineSession
@@ -22,6 +23,7 @@ import org.junit.Assert.assertFalse
 import org.junit.Assert.assertNotNull
 import org.junit.Assert.assertNull
 import org.junit.Assert.assertTrue
+import org.junit.Before
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.ArgumentMatchers.any
@@ -54,6 +56,11 @@ import org.robolectric.RobolectricTestRunner
 
 @RunWith(RobolectricTestRunner::class)
 class GeckoEngineSessionTest {
+
+    @Before
+    fun setup() {
+        ThreadUtils.sGeckoHandler = null
+    }
 
     @Test
     fun engineSessionInitialization() {
@@ -262,13 +269,22 @@ class GeckoEngineSessionTest {
 
     @Test
     fun saveState() {
-        ThreadUtils.sGeckoHandler = Handler()
-
         val engineSession = GeckoEngineSession(mock(GeckoRuntime::class.java))
         engineSession.geckoSession = mock(GeckoSession::class.java)
         val currentState = GeckoSession.SessionState("")
         val stateMap = mapOf(GeckoEngineSession.GECKO_STATE_KEY to currentState.toString())
 
+        var first = true
+        ThreadUtils.sGeckoHandler = object : Handler() {
+            override fun sendMessageAtTime(msg: Message?, uptimeMillis: Long): Boolean {
+                // Only run the saveState runnable and ignore everything that comes after
+                if (first) {
+                    first = false
+                    return super.sendMessageAtTime(msg, uptimeMillis)
+                }
+                return true
+            }
+        }
         `when`(engineSession.geckoSession.saveState()).thenReturn(GeckoResult.fromValue(currentState))
         assertEquals(stateMap, engineSession.saveState())
     }

--- a/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt
+++ b/components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt
@@ -5,6 +5,7 @@
 package mozilla.components.browser.engine.gecko
 
 import android.os.Handler
+import android.os.Message
 import mozilla.components.browser.errorpages.ErrorType
 import mozilla.components.concept.engine.DefaultSettings
 import mozilla.components.concept.engine.EngineSession
@@ -22,6 +23,7 @@ import org.junit.Assert.assertFalse
 import org.junit.Assert.assertNotNull
 import org.junit.Assert.assertNull
 import org.junit.Assert.assertTrue
+import org.junit.Before
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.ArgumentMatchers.any
@@ -54,6 +56,11 @@ import org.robolectric.RobolectricTestRunner
 
 @RunWith(RobolectricTestRunner::class)
 class GeckoEngineSessionTest {
+
+    @Before
+    fun setup() {
+        ThreadUtils.sGeckoHandler = null
+    }
 
     @Test
     fun engineSessionInitialization() {
@@ -262,13 +269,22 @@ class GeckoEngineSessionTest {
 
     @Test
     fun saveState() {
-        ThreadUtils.sGeckoHandler = Handler()
-
         val engineSession = GeckoEngineSession(mock(GeckoRuntime::class.java))
         engineSession.geckoSession = mock(GeckoSession::class.java)
         val currentState = GeckoSession.SessionState("")
         val stateMap = mapOf(GeckoEngineSession.GECKO_STATE_KEY to currentState.toString())
 
+        var first = true
+        ThreadUtils.sGeckoHandler = object : Handler() {
+            override fun sendMessageAtTime(msg: Message?, uptimeMillis: Long): Boolean {
+                // Only run the saveState runnable and ignore everything that comes after
+                if (first) {
+                    first = false
+                    return super.sendMessageAtTime(msg, uptimeMillis)
+                }
+                return true
+            }
+        }
         `when`(engineSession.geckoSession.saveState()).thenReturn(GeckoResult.fromValue(currentState))
         assertEquals(stateMap, engineSession.saveState())
     }

